open-legends.lua

```diff
@@ -1,82 +1,44 @@
 -- open legends screen when in fortress mode
 --@ module = true
---[====[
 
-open-legends
-============
-Open a legends screen when in fortress mode. Requires a world loaded in fortress
-or adventure mode. Compatible with `exportlegends`.
-
-Note that this script carries a significant risk of save corruption if the game
-is saved after exiting legends mode. To avoid this:
-
-1. Pause DF
-2. Run `quicksave` to save the game
-3. Run `open-legends` (this script) and browse legends mode as usual
-4. Immediately after exiting legends mode, run `die` to quit DF without saving
-   (saving at this point instead may corrupt your save)
-
-Note that it should be safe to run "open-legends" itself multiple times in the
-same DF session, as long as DF is killed immediately after the last run.
-Unpausing DF or running other commands risks accidentally autosaving the game,
-which can lead to save corruption.
-
-The optional ``force`` argument will bypass all safety checks, as well as the
-save corruption warning.
-
-]====]
-
-local dialogs = require 'gui.dialogs'
-local gui = require 'gui'
-local utils = require 'utils'
-
-Wrapper = defclass(Wrapper, gui.Screen)
-Wrapper.focus_path = 'legends'
-
-local region_details_backup = {} --as:df.world_region_details[]
-
-function Wrapper:onRender()
-    self._native.parent:render()
-end
-
-function Wrapper:onIdle()
-    self._native.parent:logic()
+local dialogs = require('gui.dialogs')
+local gui = require('gui')
+local utils = require('utils')
+
+Restorer = defclass(Restorer, gui.Screen)
+Restorer.ATTRS{
+    focus_path='open-legends'
+}
+
+function Restorer:init()
+    print('initializing restorer')
+    self.region_details_backup = {} --as:df.world_region_details[]
+    local v = df.global.world.world_data.region_details
+    while (#v > 0) do
+        table.insert(self.region_details_backup, 1, v[0])
+        v:erase(0)
+    end
 end
 
-function Wrapper:onHelp()
-    self._native.parent:help()
+function Restorer:onIdle()
+    self:dismiss()
 end
 
-function Wrapper:onInput(keys)
-    if self._native.parent.cur_page == 0 and keys.LEAVESCREEN then --hint:df.viewscreen_legendsst
-        local v = df.global.world.world_data.region_details
-        while (#v > 0) do v:erase(0) end
-        for _,item in pairs(region_details_backup) do
-            v:insert(0, item)
-        end
-        self:dismiss()
-        dfhack.screen.dismiss(self._native.parent)
```

Comment:
Per a discussion with Rumrusher on Discord: this dismiss() call may really be necessary, because DF's dismiss logic for this screen kicks the user back to the main menu. In previous versions, I recall it crashing instead - that's why the dismiss() call was added here.

Side note: that's why I named the class "Wrapper" instead of "Restorer". It does more than just restore the original screen/state: it also catches and overrides some native behavior to address issues like these.

Reply:
The problem is that LEAVESCREEN is no longer the only way to exit, and the mouse-driven way is not easy to detect. This is why I rewrote the logic to not care how the screen was exited. the script crashed both before and after my changes, though, so I left the tags as "unavailable" and deferred research for a later time.

Reply:
note we *may* be able to use a tactic similar to the `confirm` plugin for detecting (and intercepting) mouse clicks over the `Done` button.

Reply:
Yeah, that is tricky. The logic in viewscreen_legendsst that dismisses the screen is probably the same logic that kicks the user back to the main menu, so we probably can't just detect whether the screen's dismiss flag is set after user input is handled...

Adding detection for button clicks is probably more reliable, but tricky to make reliable, e.g. if it depends on pixel coordinates and we are off by one relative to DF.